modules/biobambam/bamsormadup/main.nf

```diff
@@ -0,0 +1,38 @@
+process BIOBAMBAM_BAMSORMADUP {
+    tag "$meta.id"
+    label 'process_medium'
+
+    conda (params.enable_conda ? "bioconda::biobambam=2.0.183" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/biobambam:2.0.183--h9f5acd7_1':
+        'quay.io/biocontainers/biobambam:2.0.183--h9f5acd7_1' }"
+
+    input:
+    tuple val(meta), path(bam)
+
+    output:
+    tuple val(meta), path("*.bam")        , emit: bam
+    tuple val(meta), path("*.metrics.txt"), emit: metrics
+    path "versions.yml"                   , emit: versions
```

Comment:
Ok, the help message is a bit difficult to understand for this tool.

Please make sure that all possible input and output channels are listed here [^1]. For example I guess `indexfilename` and `reference` could be possibles?

Optional input channels can be passed `[]` if not needed by the pipeline developer (andin tests)
For optional output files, you can specify `,optional: true` before the `emit`. 

Don't forget to update `meta.yml` accordingly

[^1]: https://nf-co.re/developers/modules#inputoutput-options points 1-3

Reply:
Do we want to add all of the options? Or can we make some assumptions and leave out some stuff?
If we want everything we'd have to make sure there's support for:
- using a real TMP dir (e.g. scratch) for TMP, instead of the PWD
- add support for a digest with a completely variable filename
- add support form cram output (in conjunction with the reference param)

Reply:
Basically we don't want to make (in most cases) assumptions what someone would want to do for a given tool. So maybe they want the extra files that the tool needs.

1. In this case it's OK to ignore - the nextflow `work/` is technically a scratch, and presumably(?) a user would never want to access these scratch files?
2. Not sure what you mean here
3. Yes, correct - `    tuple val(meta), path("*.{bam,cram}")        , emit: bam` - there are other modules you can look at to find how this is typically implemented

Reply:
1. Ok, that checks out.
2. Bamsormadup can output a file checksum with different algorithms, the name for this file can be completely random. How should we handle this? 
3. Ok, should be fixed

Reply:
> Bamsormadup can output a file checksum with different algorithms, the name for this file can be completely random. How should we handle this?

Ooff, ok that is a tricky one... I guess we can sweep that under the rug if it's completely random...

Reply:
consider it swept! I don't think that feature sees much use TBH